File: Week_02-哈希表-集合-树-图/102.二叉树的层序遍历.py
#
# @lc app=leetcode.cn id=102 lang=python
#
# [102] 二叉树的层序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        results = list()
        if not root:
            return results
        queue = collections.deque([root])
        while queue:
            size = len(queue)
            result = list()
            for _ in range(size):
                cur = queue.popleft()
                result.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
            results.append(result)
        return results
# @lc code=end

File: Week_02-哈希表-集合-树-图/102.二叉树的层序遍历.cpp
/*
 * @lc app=leetcode.cn id=102 lang=cpp
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//队列暂存下层信息，利用其先进先出特点
#include<vector>
using namespace std;

class Solution {
#if 0
/*
BFS,用队列实现。根节点放入队列，在队列取出根节点的同时，把根节点的子节点分别放入下一层；
直到队列中当前层根节点被完全访问，再访问队列中下一层根节点
*/
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        if(root) q.push(root);
        while (!q.empty()) {
            vector<int> vec;
            for (int i = q.size() - 1; i >= 0; --i) {
                TreeNode* tmp = q.front(); q.pop();
                vec.emplace_back(tmp->val);
                if(tmp->left) q.push(tmp->left);
                if(tmp->right) q.push(tmp->right);
            }
            ans.emplace_back(vec);
        }
        return ans;
    }
#endif
#if 1
/*
DFS，
*/
public:
    vector<vector<int>> res;
    vector<vector<int>> levelOrder(TreeNode* root) {
        addVector(root, 0);
        return res;
    }
    void addVector(TreeNode* root, int level) {
        if(root == NULL) return;
        if(res.size() == level) res.resize(level+1);//第一次访问当前层（不是return回到当前层），那就在数组中扩充放置下一层数据的容量
        res[level].push_back(root->val);
        addVector(root->left, level+1);
        addVector(root->right, level+1);
    }
#endif
};
// @lc code=end

